[Low Kang Ngee] iP

lowkangn opened the pull request:
### DukeNukemBot
> "It's time to kick ass and chew bubble gum... and I'm all outta gum."
> – Duke Nukem

DukeNukemBot, Duke for short, helps you do the following things:
- Add tasks such as _todos_, _events_ and _deadlines_ :memo:
- Mark tasks as done :heavy_check_mark:
- Delete tasks :x:

All you need to do is:
1. Download it from [here](https://github.com/lowkangn/ip/releases/tag/A-Gradle).
2. Copy duke.jar into an empty folder.
3. Open a command window in the folder.
4. Run the program using the command `java -jar duke.jar`
5. Start using Duke!

Super **FAST** and **EASY** to use! ~~(If you're good at typing, that is)~~

Additional features coming soon:
- [ ] Managing deadlines
- [ ] Reminders


If you are a Java programmer, you can use it to practice Java too! Here's the `main` method:
```ruby
public static void main(String[] args) {
        Duke duke = new Duke("save.txt");
        duke.run();
    }
```

Commit:
Add Gradle support

lowkangn committed:
Skeletal version of Duke, echos commands entered

lowkangn committed:
Added ability to store user inputs and display them back to the user

lowkangn committed:
Added ability to mark tasks as done

lowkangn committed:
Added support for tracking todos, deadlines and events

lowkangn committed:
Semi-automated testing of Duke

lowkangn committed:
Added error handling to Duke

lowkangn committed:
Added support for deleting tasks from the list

lowkangn committed:
Add enums for tasks

lowkangn committed:
Add ability to save and load task list

lowkangn committed:
Update save file after each action rather than only on exit

lowkangn committed:
Add support for dates and times

lowkangn committed:
Add support for end time for events

lowkangn committed:
Refactor code by extracting classes

lowkangn committed:
Divide classes into packages

lowkangn committed:
Write junit tests for TaskList

lowkangn committed:
Create JAR file

lowkangn committed:
Add and edit JavaDoc comments

lowkangn committed:
Tweak code to follow Java coding standard

lowkangn committed:
Add ability to find tasks by searching for a keyword

lowkangn committed:
Merge branch branch-A-JavaDoc

lowkangn committed:
Merge tag 'A-JavaDoc'

Add and edit JavaDoc comments.

lowkangn committed:
Merge tag 'A-CodingStandard'

Tweak code to follow Java coding standard.

lowkangn committed:
Merge branch branch-Level-9 into master

lowkangn committed:
Add gradle support

lowkangn committed:
Add Gradle support

lowkangn committed:
Edit to fit Java coding standard

lowkangn committed:
Add checks for CheckStyle

jason-ng-zq99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+    public static Task processTaskString(String taskString) {
+        char type = taskString.charAt(1);
+        boolean isDone = taskString.charAt(0) == 'X';
+        String[] split = taskString.split("/");
```
I think it will be less confusing to name it splitString

jason-ng-zq99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,180 @@
+                            break;
+
+                        default:
+                            String[] nameAndTime = substring.split(" /at ");
```
Great and clear naming of variables! Quite consistent throughout your code :smile:

jason-ng-zq99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     */
+    public static LocalDateTime formatDateTime(String dateAndTime) throws DukeException {
+        try {
+            String[] split = dateAndTime.split(", ");
```
Similarly, I think it will be less confusing to name it something more than `split`. May I suggest `splitAtComma`?

jason-ng-zq99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,78 @@
+     * @throws DukeException If there was an error reading or writing to the save file.
+     */
+    public ArrayList<Task> load() throws DukeException {
+        File save = new File(this.filePath);
```
Perhaps a clearer name like ```saveFile``` would be better! :smile:

jason-ng-zq99 reviewed (commented):
Overall, I think you are all good except really minor naming issues which are inevitable in a project! Good luck to your development in Duke. :smile:

lowkangn committed:
Add GUI to Duke

lowkangn committed:
Edit to restore Duke's original functionality

lowkangn committed:
Edit to restore Duke's original functionality

lowkangn committed:
Edit to restore Duke's original functionality

lowkangn committed:
Fix for JAR build

NaokiShum reviewed (commented):
Dont think im able to spot any blatant coding standard violations. Variable/method names also closely follows the code quality guideline. Overall code looks very good and neat. All the best for the rest of the project !

lowkangn committed:
Fix for JAR build

The duke.jar file built by Gradle does not make use of the GUI.
The JAR file requires a command window to properly run Duke.

Let's adjust the manifest file so that Gradle builds the JAR
file correctly.

lowkangn committed:
Add assertion to Storage and Parser

Storage must ensure that it receives a valid TaskList before
saving to the save file. Parser must receive valid String
inputs to convert them to Task objects correctly.

Let's add assertions to assert that the inputs passed into the
relevant methods are correctly configured.

Assertions are useful to define assumptions about the program
state so that the runtime can verify them.

lowkangn committed:
Update code to improve code quality

Complicated boolean expressions in the code reduce readability.
Some comments repeat the obvious.

Let's extract the complicated boolean expressions into boolean
variables to improve readability of the code. Remove or improve
unnecessary comments.

Code quality is important to help the code be more understandable
and easier modified by other developers or by oneself.

lowkangn committed:
Merge pull request #1 from lowkangn/branch-A-Assertions

Merge branch-A-Assertions

lowkangn committed:
Merge branch 'master' into branch-A-CodeQuality

lowkangn committed:
Merge pull request #2 from lowkangn/branch-A-CodeQuality

Branch a code quality

lowkangn committed:
Add ability to update tasks

To update any task in the task list, the task has to be deleted
and a new task created. Not only is it tedious, the task's
position in the list is also not preserved.

Let's add a new update command to allow the user to update
existing tasks without having to delete them. The user will
be able to change the description, time or both of a task.

With this approach, the task's position in the list will
not be affected.

lowkangn committed:
Add personality to Duke

Duke's current replies are very generic and do not reflect
Duke Nukem's mannerisms.

Duke's replies to the user are replaced with ones that are
more true to how he would talk. (within reason!)

lowkangn committed:
Update README.md

README does not provide user guide for using Duke.

New users of Duke will have no clue on how to use Duke properly.

Let's update the README.md to provide a user guide for users to
learn the commands to use Duke.

lowkangn committed:
Finalise Duke for release

Javadoc comment for main method adjusted to be more descriptive.

lowkangn committed:
Edit README.md

Let's fix the errors present in the user guide in README.md.

lowkangn committed:
Set theme jekyll-theme-cayman

lowkangn committed:
Set theme jekyll-theme-cayman

lowkangn committed:
Set theme jekyll-theme-cayman
